File: homework4/fortran/pingPong.f90
PROGRAM pingPong
INCLUDE 'mpif.h'

INTEGER :: ierr, rank,nNodes,i
INTEGER :: bounce = 1
INTEGER :: source = 0
INTEGER :: destination = 1
INTEGER :: stat(MPI_STATUS_SIZE)

!initialize MPI
CALL MPI_INIT(ierr);
!get number of processors
CALL MPI_COMM_SIZE(MPI_COMM_WORLD, nNodes,ierr)
!get the rank of the processor
CALL MPI_COMM_RANK(MPI_COMM_WORLD,rank,ierr)

IF(rank .EQ. source) THEN
    bounce = 5
END IF

DO WHILE(0 < bounce )
    IF(rank .EQ. source) THEN
        CALL MPI_SEND(bounce,1,MPI_INTEGER,destination,0,MPI_COMM_WORLD,ierr)
        print *,"message sent: ",bounce," from processor ", rank, " to processor ", destination 
        
    ELSE IF(rank .EQ. destination) THEN
        CALL MPI_RECV(bounce,1,MPI_INTEGER,source,0,MPI_COMM_WORLD,stat,ierr)
        print *,"message received: ",bounce," on processor ", rank , "from processor ", source
    END IF
    source = ieor(source,1)
    destination = ieor(destination,1)
    bounce = bounce-1

END DO

CALL MPI_Finalize(ierr);

STOP
END PROGRAM pingPong

File: homework4/turnIn/pingPong/pingPong.c
#include <stdio.h>
#include <mpi.h>
#include <stdlib.h>

main(int argc, char** argv){
    int ierr;
    int nNodes;
    int rank;
    int destination = 1;
    int temp;
    int source = 0;
    int bounces = 5;

    ierr = MPI_Init(NULL, NULL);
    //get number of total processors
    MPI_Comm_size(MPI_COMM_WORLD, &nNodes);
    //get rank of processor
    MPI_Comm_rank(MPI_COMM_WORLD,&rank);

     if (nNodes < 2) {
        fprintf(stderr, "World size must be greater than 1 for %s\n", argv[0]);
        MPI_Abort(MPI_COMM_WORLD, 1);
    }

    //if it is the source processor, send
    while  (0 < bounces){
        if(rank == source){
            
            MPI_Send(&bounces,1,MPI_INT, destination,0,MPI_COMM_WORLD);
            printf("Processor %d sent the message: '%d bounces left'  to processor %d\n", rank, bounces, destination);
            

        //if it is the destination processor, receive
        }else if (rank==destination) {
            MPI_Recv(&bounces,1,MPI_INT, source,0,MPI_COMM_WORLD,MPI_STATUS_IGNORE);
            printf("Processor %d received the message: '%d bounces left' from processor %d\n", rank, bounces, source);

        }
        printf("\n");
        source = !source;
        destination = !destination;
        bounces--;
    }
    ierr = MPI_Finalize();

}
